drivers/sensor/akm09918c/akm09918c.c

```diff
@@ -204,6 +223,189 @@ static int akm09918c_init(const struct device *dev)
 	return 0;
 }
 
+/*
+ * RTIO Sensor V2 API implementation
+ */
+
+static int akm09918c_encode(const struct device *dev, const enum sensor_channel *const channels,
+			    const size_t num_channels, uint8_t *buf)
+{
+	/* This sensor only supports SENSOR_CHAN_MAGN_XYZ */
+	ARG_UNUSED(channels);
+	ARG_UNUSED(num_channels);
+
+	struct akm09918c_encoded_data *edata = (struct akm09918c_encoded_data *)buf;
+
+	/* Only one-shot mode supported at this time. */
+	edata->header.is_fifo = false;
+	edata->header.timestamp = k_ticks_to_ns_floor64(k_uptime_ticks());
+
+	return 0;
+}
+
+static int akm09918c_submit(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
+{
+	const struct sensor_read_config *cfg = iodev_sqe->sqe.iodev->data;
+	const enum sensor_channel *const channels = cfg->channels;
+	const size_t num_channels = cfg->count;
+	uint32_t min_buf_len = sizeof(struct akm09918c_encoded_data);
+	int rc;
+	uint8_t *buf;
+	uint32_t buf_len;
+	struct akm09918c_encoded_data *edata;
+
+	/* Get the buffer for the frame, it may be allocated dynamically by the rtio context */
+	rc = rtio_sqe_rx_buf(iodev_sqe, min_buf_len, min_buf_len, &buf, &buf_len);
+	if (rc != 0) {
+		LOG_ERR("Failed to get a read buffer of size %u bytes", min_buf_len);
+		rtio_iodev_sqe_err(iodev_sqe, rc);
+		return rc;
+	}
+
+	edata = (struct akm09918c_encoded_data *)buf;
+
+	akm09918c_encode(dev, channels, num_channels, buf);
+
+	rc = akm09918c_sample_fetch_helper(dev, SENSOR_CHAN_MAGN_XYZ, &edata->readings[0],
+					   &edata->readings[1], &edata->readings[2]);
+	if (rc != 0) {
+		LOG_ERR("Failed to fetch samples");
+		rtio_iodev_sqe_err(iodev_sqe, rc);
+		return rc;
+	}
+
+	rtio_iodev_sqe_ok(iodev_sqe, 0);
+
+	return 0;
+}
+
+/*
+ * RTIO Decoder API
+ */
+
+static int akm09918c_decoder_get_frame_count(const uint8_t *buffer, enum sensor_channel channel,
+					     size_t channel_idx, uint16_t *frame_count)
+{
+	const struct akm09918c_encoded_data *data = (const struct akm09918c_encoded_data *)buffer;
+	const struct akm09918c_decoder_header *header = &data->header;
+
+	if (!header->is_fifo) {
+		*frame_count = 1;
+		return 0;
+	}
+
+	/* Only one-shot mode is currently supported */
+	return -ENOTSUP;
+}
+
+static int akm09918c_decoder_get_size_info(enum sensor_channel channel, size_t *base_size,
+					   size_t *frame_size)
+{
+	switch (channel) {
+	case SENSOR_CHAN_MAGN_X:
+	case SENSOR_CHAN_MAGN_Y:
+	case SENSOR_CHAN_MAGN_Z:
+	case SENSOR_CHAN_MAGN_XYZ:
+		*base_size = sizeof(struct sensor_three_axis_data);
+		*frame_size = sizeof(struct sensor_three_axis_sample_data);
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+/** Fixed shift value to use. All channels (MAGN_X, _Y, and _Z) have the same fixed range of
+ *  +/- 49.12 Gauss.
+ */
+#define AKM09918C_SHIFT (6)
+
+static int akm09918c_convert_raw_to_q31(int16_t reading, q31_t *out)
+{
+	/* Convert and store reading with separate whole and fractional parts */
+	struct sensor_value val;
+
+	akm09918c_convert(&val, reading);
```

Comment:
Can we do this without first converting to `sensor_value`? It'll eventually be deprecated entirely and adds more rounding error.